bottlerocket/agents/src/bin/ec2-resource-agent/ec2_provider.rs

```diff
@@ -120,70 +120,73 @@ impl Create for Ec2Creator {
 
         // Determine the instance type to use. If provided use that one. Otherwise, for `x86_64` use `m5.large`
         // and for `aarch64` use `m6g.large`
-        let instance_type = if let Some(instance_type) = spec.configuration.instance_type {
-            instance_type
+        let instance_types = if !spec.configuration.instance_types.is_empty() {
+            spec.configuration.instance_types
         } else {
-            instance_type(&ec2_client, &spec.configuration.node_ami, &memo).await?
+            vec![instance_type(&ec2_client, &spec.configuration.node_ami, &memo).await?]
         };
-        info!("Using instance type '{}'", instance_type);
 
-        // Run the ec2 instances
         let instance_count = spec
             .configuration
             .instance_count
             .unwrap_or(DEFAULT_INSTANCE_COUNT);
         info!("Creating {} instance(s)", instance_count);
 
         let mut run_instance_result = None;
-        for subnet_id in &spec.configuration.subnet_ids {
-            let run_instances = ec2_client
-                .run_instances()
-                .min_count(instance_count)
-                .max_count(instance_count)
-                .subnet_id(subnet_id)
-                .set_security_group_ids(Some(spec.configuration.security_groups.clone()))
-                .image_id(&spec.configuration.node_ami)
-                .instance_type(InstanceType::from(instance_type.as_str()))
-                .tag_specifications(tag_specifications(
-                    &spec.configuration.cluster_type,
-                    &spec.configuration.cluster_name,
-                    &instance_uuid,
-                ))
-                .user_data(userdata(
-                    &spec.configuration.cluster_type,
-                    &spec.configuration.cluster_name,
-                    &spec.configuration.endpoint,
-                    &spec.configuration.certificate,
-                    &spec.configuration.cluster_dns_ip,
-                    &memo,
-                )?)
-                .iam_instance_profile(
-                    IamInstanceProfileSpecification::builder()
-                        .arn(&spec.configuration.instance_profile_arn)
-                        .build(),
-                );
+        'outer: for instance_type in instance_types {
+            info!("Using instance type '{}'", instance_type);
```

Comment:
Maybe we can refactor this into a function and just return  the `instance_result` when we've successfully launched the instances instead of breaking loops awkwardly here.

Reply:
I agree. The "inner loop" (for subnet id) could be done in a function which would improve readability. @ecpullen why do we loop over subnet ids?

Reply:
I'll put that inner loop in a different function. Some context for why we loop over subnet ids is in this PR: https://github.com/bottlerocket-os/bottlerocket-test-system/pull/585 and this issue: https://github.com/bottlerocket-os/bottlerocket-test-system/issues/578 . Essentially, it was possible that an instance type was not available on the randomly selected availability zone but it might be available on another one, so we try to create the instance on each AZ until one works.

Reply:
One possible issue with what we are doing here... we aren't checking the reason for the failure. So if I understand correctly, we might receive an error such as "Not Authorized" and we would go through all of this looping anyway, right? Or are we checking that the issue is actually a lack of capacity?